l2-contracts/contracts/ConsensusRegistry.sol

```diff
@@ -0,0 +1,278 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.20;
+
+import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";
+import {EfficientCall} from "@matterlabs/zksync-contracts/l2/system-contracts/libraries/EfficientCall.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @title ConsensusRegistry
+/// @dev Manages consensus nodes and committees for the L2 consensus protocol,
+/// owned by a governance protocol. Nodes act as both validators and attesters,
+/// each playing a distinct role in the consensus process. This contract facilitates
+/// the rotation of validator and attester committees, which represent a subset of nodes
+/// expected to actively participate in the consensus process during a specific time window.
+contract ConsensusRegistry is Ownable2Step {
+    // An array to keep track of node owners.
+    address[] public nodeOwners;
+    // A map of node owners => nodes.
+    mapping(address => Node) public nodes;
+    // The current validator committee list.
+    CommitteeValidator[] public validatorCommittee;
+    // The current attester committee list.
+    CommitteeAttester[] public attesterCommittee;
+
+    /// @dev Represents a consensus node.
+    struct Node {
+        // A flag stating if the node is active.
+        // Inactive nodes are not considered when selecting committees.
+        bool active;
+        // Validator's voting weight.
+        uint256 validatorWeight;
+        // Validator's BLS12-381 public key.
+        bytes validatorPubKey;
+        // Validator's Proof-of-possession (a signature over the public key).
+        bytes validatorPoP;
+        // Attester's Voting weight.
+        uint256 attesterWeight;
+        // Attester's ECDSA public key.
+        bytes attesterPubKey;
+    }
+
+    /// @dev Represents a validator committee member.
+    struct CommitteeValidator {
+        address nodeOwner;
+        uint256 weight;
+        bytes pubKey;
+        bytes pop;
+    }
+
+    /// @dev Represents an attester committee member.
+    struct CommitteeAttester {
+        uint256 weight;
+        address nodeOwner;
+        bytes pubKey;
+    }
+
+    error UnauthorizedOnlyOwnerOrNodeOwner();
+    error NodeOwnerAlreadyExists();
+    error NodeOwnerDoesNotExist();
+    error NodeOwnerNotFound();
+    error ValidatorPubKeyAlreadyExists();
+    error AttesterPubKeyAlreadyExists();
+
+    modifier onlyOwnerOrNodeOwner(address _nodeOwner) {
+        address sender = _msgSender();
```

Comment:
`_msgSender` function is designed only to support relayed transactions, you can use `msg.sender` for simplicity instead as we never use relayed txs.

Reply:
Fixed here: https://github.com/matter-labs/era-contracts/pull/555/commits/8c5427948a13b796d44398e61e20a92a93fac5bb